torch/_logging.py

```diff
@@ -0,0 +1,286 @@
+        clean_name = clean_name.replace(DECR_VERBOSITY_CHAR, "")
+        level = None
+        if clean_name in VERBOSE_NAMES:
+            if name[0] == INCR_VERBOSITY_CHAR:
+                level = logging.DEBUG
+            elif name[0] == DECR_VERBOSITY_CHAR:
+                level = logging.ERROR
+            else:
+                level = logging.INFO
+
+        return clean_name, level
+
+    name_levels = [get_name_level_pair(name) for name in log_names]
+    return {name: level for name, level in name_levels}
+
+
+class FilterByType(logging.Filter):
+    def __init__(self, enabled_types):
+        self.set_enabled_types(enabled_types)
+
+    def filter(self, record):
+        return isinstance(record.msg, self.enabled_types)
+
+    def set_enabled_types(self, enabled_types):
+        self.enabled_types = tuple(set(enabled_types))
+
+
+def _get_log_settings():
+    log_setting = os.environ.get("TORCH_LOGS", None)
+    if log_setting is None:
+        return {}
+    else:
+        return _parse_log_settings(log_setting)
+
+
+# setup custom handlers
+# if the log level of a component is set to INFO, setup
+# an additional handler to print those messages, because
+# the debug handler is what handles custom objects like guards,
+# bytecode, etc.
+# if the log level of a component is set to DEBUG, allow all
+# string messages and allowed types (other than those off by default)
+def _setup_handlers(create_handler_fn, log, enabled_types, formatter, level=None):
+    debug_handler = create_handler_fn()
+    debug_handler.setFormatter(formatter)
+    debug_handler.setLevel(logging.DEBUG)
+
+    if level == logging.DEBUG:
+        enabled_types = enabled_types.union({str})
+
+    filter = FilterByType(enabled_types)
+    debug_handler.addFilter(filter)
+    log.addHandler(debug_handler)
+
+    if level is not None and level > logging.DEBUG:
+        generic_handler = create_handler_fn()
+        generic_handler.setFormatter(formatter)
+        generic_handler.setLevel(level)
+        log.addHandler(generic_handler)
+
+
+# mark handlers that we've created
+# so we don't modify user handlers
+def _tag_handler(handler):
+    handler.__torch_log_handler = True
+    return handler
+
+
+def _is_torch_handler(handler):
+    return hasattr(handler, "__torch_log_handler")
+
+
+# clears all torch handlers on specified loggers
+def _clear_handlers(log):
+    to_remove = [handler for handler in log.handlers if _is_torch_handler(handler)]
+    for handler in to_remove:
+        log.removeHandler(handler)
+
+
+# initialize loggers log_names
+# each developer component should call this for their own logs
+# in the appropriate location after relevant types have been registered
+def init_logs(log_names, log_file_name=None, formatter=None):
+    if not formatter:
+        formatter = DEFAULT_FORMATTER
+
+    for log_name in log_names:
+        log = logging.getLogger(log_name)
+        log.setLevel(logging.DEBUG)  # allow all messages through to the handlers
```

Comment:
Does this have performance implications? If you set the level on the component logger to debug, we will ALWAYS generate log messages for all debug unless they all get discarded. There's a strong performance impetus to avoid building log messages as *early* as possible

Reply:
Messages aren't rendered until they are sent to a handler which in this design has a filter which accepts/rejects records before the __str__ method is called

Reply:
Sometimes loggers will test the level on the logger to see if they should send a log message or not (because they are too lazy to use the built in lazy formatting in logging to construct message). So that seems bad for this case.

Reply:
Discussed offline, we may use a linter to enforce this, but not sure if there is a perfect solution for this, tbh.

Reply:
I will do this linter in a follow up PR